Resolve import issues and update playwright tests

Current attempt to run tests on `dev` result in the following:

```
tests/test_dashboard_playwright.py::test_initial_kpi_count[chromium] ERROR                                                                                [ 12%]
tests/test_dashboard_playwright.py::test_customer_table_initial_structure[chromium] ERROR                                                                 [ 25%]
tests/test_dashboard_playwright.py::test_customer_table_initial_cell_values[chromium] ERROR                                                               [ 37%]
tests/test_dashboard_playwright.py::test_region_filter_asia_only[chromium] ERROR                                                                          [ 50%]
tests/test_dashboard_playwright.py::test_purchase_type_filter_two_values[chromium] ERROR                                                                  [ 62%]
tests/test_dashboard_playwright.py::test_retention_strategy_filter_two_values[chromium] ERROR                                                             [ 75%]
tests/test_dashboard_playwright.py::test_row_dropdown_changes_grouping[chromium] ERROR                                                                    [ 87%]
tests/test_dashboard_playwright.py::test_reset_button_restores_defaults[chromium] ERROR                                                                   [100%]

============================================================================ ERRORS =============================================================================
______________________________________________________ ERROR at setup of test_initial_kpi_count[chromium] _______________________________________________________

request = <SubRequest 'app' for <Function test_initial_kpi_count[chromium]>>

    @pytest.fixture(scope=scope)
    def fixture_func(request: pytest.FixtureRequest):
        app_path = get_app_path(request, app)
        sa_gen = shiny_app_gen(app_path, timeout_secs=timeout_secs)
>       yield next(sa_gen)
              ^^^^^^^^^^^^

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/pytest/_fixture.py:149: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:341: in shiny_app_gen
    sa = run_shiny_app(
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:269: in run_shiny_app
    sa.wait_until_ready(timeout_secs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <shiny.run._run.ShinyAppProc object at 0x111fe78f0>, timeout_secs = 30

    def wait_until_ready(self, timeout_secs: float) -> None:
        """
        Waits until the shiny app is ready to serve requests.
    
        Parameters
        ----------
        timeout_secs
            The maximum number of seconds to wait for the app to become ready.
    
        Raises
        ------
        ConnectionError
            If there is an error while starting the shiny app.
        TimeoutError
            If the shiny app does not become ready within the specified timeout.
        """
        error_lines: List[str] = []
    
        def stderr_uvicorn(line: str) -> bool:
            error_lines.append(line)
            if "error while attempting to bind on address" in line:
                raise ConnectionError(f"Error while starting shiny app: `{line}`")
            return "Uvicorn running on" in line
    
        if self.stderr.wait_for(stderr_uvicorn, timeout_secs=timeout_secs):
            return
        else:
>           raise TimeoutError(
                "Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:\n"
                + "\n".join(error_lines[-20:])
            )
E           TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/server.py", line 78, in _serve
E           
E               config.load()
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/config.py", line 436, in load
E           
E               self.loaded_app = import_from_string(self.app)
E           
E                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/importer.py", line 19, in import_from_string
E           
E               module = importlib.import_module(module_str)
E           
E                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/importlib/__init__.py", line 90, in import_module
E           
E               return _bootstrap._gcd_import(name[level:], package, level)
E           
E                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
E           
E             File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
E           
E             File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
E           
E             File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
E           
E             File "<frozen importlib._bootstrap_external>", line 999, in exec_module
E           
E             File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
E           
E             File "/Users/alxti/Documents/alex stuff/UBC/block5/532/DSCI-532_2026_10_Salescope/src/app.py", line 17, in <module>
E           
E               from .dflogic import create_summary_table, filter_sales_data
E           
E           ImportError: attempted relative import with no known parent package

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:198: TimeoutError
_______________________________________________ ERROR at setup of test_customer_table_initial_structure[chromium] _______________________________________________

request = <SubRequest 'app' for <Function test_initial_kpi_count[chromium]>>

    @pytest.fixture(scope=scope)
    def fixture_func(request: pytest.FixtureRequest):
        app_path = get_app_path(request, app)
        sa_gen = shiny_app_gen(app_path, timeout_secs=timeout_secs)
>       yield next(sa_gen)
              ^^^^^^^^^^^^

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/pytest/_fixture.py:149: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:341: in shiny_app_gen
    sa = run_shiny_app(
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:269: in run_shiny_app
    sa.wait_until_ready(timeout_secs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <shiny.run._run.ShinyAppProc object at 0x111fe78f0>, timeout_secs = 30

    def wait_until_ready(self, timeout_secs: float) -> None:
        """
        Waits until the shiny app is ready to serve requests.
    
        Parameters
        ----------
        timeout_secs
            The maximum number of seconds to wait for the app to become ready.
    
        Raises
        ------
        ConnectionError
            If there is an error while starting the shiny app.
        TimeoutError
            If the shiny app does not become ready within the specified timeout.
        """
        error_lines: List[str] = []
    
        def stderr_uvicorn(line: str) -> bool:
            error_lines.append(line)
            if "error while attempting to bind on address" in line:
                raise ConnectionError(f"Error while starting shiny app: `{line}`")
            return "Uvicorn running on" in line
    
        if self.stderr.wait_for(stderr_uvicorn, timeout_secs=timeout_secs):
            return
        else:
>           raise TimeoutError(
                "Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:\n"
                + "\n".join(error_lines[-20:])
            )
E           TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/server.py", line 78, in _serve
E           
E               config.load()
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/config.py", line 436, in load
E           
E               self.loaded_app = import_from_string(self.app)
E           
E                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/importer.py", line 19, in import_from_string
E           
E               module = importlib.import_module(module_str)
E           
E                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/importlib/__init__.py", line 90, in import_module
E           
E               return _bootstrap._gcd_import(name[level:], package, level)
E           
E                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
E           
E             File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
E           
E             File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
E           
E             File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
E           
E             File "<frozen importlib._bootstrap_external>", line 999, in exec_module
E           
E             File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
E           
E             File "/Users/alxti/Documents/alex stuff/UBC/block5/532/DSCI-532_2026_10_Salescope/src/app.py", line 17, in <module>
E           
E               from .dflogic import create_summary_table, filter_sales_data
E           
E           ImportError: attempted relative import with no known parent package

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:198: TimeoutError
______________________________________________ ERROR at setup of test_customer_table_initial_cell_values[chromium] ______________________________________________

request = <SubRequest 'app' for <Function test_initial_kpi_count[chromium]>>

    @pytest.fixture(scope=scope)
    def fixture_func(request: pytest.FixtureRequest):
        app_path = get_app_path(request, app)
        sa_gen = shiny_app_gen(app_path, timeout_secs=timeout_secs)
>       yield next(sa_gen)
              ^^^^^^^^^^^^

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/pytest/_fixture.py:149: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:341: in shiny_app_gen
    sa = run_shiny_app(
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:269: in run_shiny_app
    sa.wait_until_ready(timeout_secs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <shiny.run._run.ShinyAppProc object at 0x111fe78f0>, timeout_secs = 30

    def wait_until_ready(self, timeout_secs: float) -> None:
        """
        Waits until the shiny app is ready to serve requests.
    
        Parameters
        ----------
        timeout_secs
            The maximum number of seconds to wait for the app to become ready.
    
        Raises
        ------
        ConnectionError
            If there is an error while starting the shiny app.
        TimeoutError
            If the shiny app does not become ready within the specified timeout.
        """
        error_lines: List[str] = []
    
        def stderr_uvicorn(line: str) -> bool:
            error_lines.append(line)
            if "error while attempting to bind on address" in line:
                raise ConnectionError(f"Error while starting shiny app: `{line}`")
            return "Uvicorn running on" in line
    
        if self.stderr.wait_for(stderr_uvicorn, timeout_secs=timeout_secs):
            return
        else:
>           raise TimeoutError(
                "Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:\n"
                + "\n".join(error_lines[-20:])
            )
E           TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/server.py", line 78, in _serve
E           
E               config.load()
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/config.py", line 436, in load
E           
E               self.loaded_app = import_from_string(self.app)
E           
E                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/importer.py", line 19, in import_from_string
E           
E               module = importlib.import_module(module_str)
E           
E                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/importlib/__init__.py", line 90, in import_module
E           
E               return _bootstrap._gcd_import(name[level:], package, level)
E           
E                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
E           
E             File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
E           
E             File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
E           
E             File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
E           
E             File "<frozen importlib._bootstrap_external>", line 999, in exec_module
E           
E             File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
E           
E             File "/Users/alxti/Documents/alex stuff/UBC/block5/532/DSCI-532_2026_10_Salescope/src/app.py", line 17, in <module>
E           
E               from .dflogic import create_summary_table, filter_sales_data
E           
E           ImportError: attempted relative import with no known parent package

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:198: TimeoutError
___________________________________________________ ERROR at setup of test_region_filter_asia_only[chromium] ____________________________________________________

request = <SubRequest 'app' for <Function test_initial_kpi_count[chromium]>>

    @pytest.fixture(scope=scope)
    def fixture_func(request: pytest.FixtureRequest):
        app_path = get_app_path(request, app)
        sa_gen = shiny_app_gen(app_path, timeout_secs=timeout_secs)
>       yield next(sa_gen)
              ^^^^^^^^^^^^

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/pytest/_fixture.py:149: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:341: in shiny_app_gen
    sa = run_shiny_app(
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:269: in run_shiny_app
    sa.wait_until_ready(timeout_secs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <shiny.run._run.ShinyAppProc object at 0x111fe78f0>, timeout_secs = 30

    def wait_until_ready(self, timeout_secs: float) -> None:
        """
        Waits until the shiny app is ready to serve requests.
    
        Parameters
        ----------
        timeout_secs
            The maximum number of seconds to wait for the app to become ready.
    
        Raises
        ------
        ConnectionError
            If there is an error while starting the shiny app.
        TimeoutError
            If the shiny app does not become ready within the specified timeout.
        """
        error_lines: List[str] = []
    
        def stderr_uvicorn(line: str) -> bool:
            error_lines.append(line)
            if "error while attempting to bind on address" in line:
                raise ConnectionError(f"Error while starting shiny app: `{line}`")
            return "Uvicorn running on" in line
    
        if self.stderr.wait_for(stderr_uvicorn, timeout_secs=timeout_secs):
            return
        else:
>           raise TimeoutError(
                "Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:\n"
                + "\n".join(error_lines[-20:])
            )
E           TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/server.py", line 78, in _serve
E           
E               config.load()
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/config.py", line 436, in load
E           
E               self.loaded_app = import_from_string(self.app)
E           
E                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/importer.py", line 19, in import_from_string
E           
E               module = importlib.import_module(module_str)
E           
E                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/importlib/__init__.py", line 90, in import_module
E           
E               return _bootstrap._gcd_import(name[level:], package, level)
E           
E                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
E           
E             File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
E           
E             File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
E           
E             File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
E           
E             File "<frozen importlib._bootstrap_external>", line 999, in exec_module
E           
E             File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
E           
E             File "/Users/alxti/Documents/alex stuff/UBC/block5/532/DSCI-532_2026_10_Salescope/src/app.py", line 17, in <module>
E           
E               from .dflogic import create_summary_table, filter_sales_data
E           
E           ImportError: attempted relative import with no known parent package

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:198: TimeoutError
_______________________________________________ ERROR at setup of test_purchase_type_filter_two_values[chromium] ________________________________________________

request = <SubRequest 'app' for <Function test_initial_kpi_count[chromium]>>

    @pytest.fixture(scope=scope)
    def fixture_func(request: pytest.FixtureRequest):
        app_path = get_app_path(request, app)
        sa_gen = shiny_app_gen(app_path, timeout_secs=timeout_secs)
>       yield next(sa_gen)
              ^^^^^^^^^^^^

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/pytest/_fixture.py:149: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:341: in shiny_app_gen
    sa = run_shiny_app(
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:269: in run_shiny_app
    sa.wait_until_ready(timeout_secs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <shiny.run._run.ShinyAppProc object at 0x111fe78f0>, timeout_secs = 30

    def wait_until_ready(self, timeout_secs: float) -> None:
        """
        Waits until the shiny app is ready to serve requests.
    
        Parameters
        ----------
        timeout_secs
            The maximum number of seconds to wait for the app to become ready.
    
        Raises
        ------
        ConnectionError
            If there is an error while starting the shiny app.
        TimeoutError
            If the shiny app does not become ready within the specified timeout.
        """
        error_lines: List[str] = []
    
        def stderr_uvicorn(line: str) -> bool:
            error_lines.append(line)
            if "error while attempting to bind on address" in line:
                raise ConnectionError(f"Error while starting shiny app: `{line}`")
            return "Uvicorn running on" in line
    
        if self.stderr.wait_for(stderr_uvicorn, timeout_secs=timeout_secs):
            return
        else:
>           raise TimeoutError(
                "Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:\n"
                + "\n".join(error_lines[-20:])
            )
E           TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/server.py", line 78, in _serve
E           
E               config.load()
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/config.py", line 436, in load
E           
E               self.loaded_app = import_from_string(self.app)
E           
E                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/importer.py", line 19, in import_from_string
E           
E               module = importlib.import_module(module_str)
E           
E                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/importlib/__init__.py", line 90, in import_module
E           
E               return _bootstrap._gcd_import(name[level:], package, level)
E           
E                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
E           
E             File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
E           
E             File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
E           
E             File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
E           
E             File "<frozen importlib._bootstrap_external>", line 999, in exec_module
E           
E             File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
E           
E             File "/Users/alxti/Documents/alex stuff/UBC/block5/532/DSCI-532_2026_10_Salescope/src/app.py", line 17, in <module>
E           
E               from .dflogic import create_summary_table, filter_sales_data
E           
E           ImportError: attempted relative import with no known parent package

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:198: TimeoutError
_____________________________________________ ERROR at setup of test_retention_strategy_filter_two_values[chromium] _____________________________________________

request = <SubRequest 'app' for <Function test_initial_kpi_count[chromium]>>

    @pytest.fixture(scope=scope)
    def fixture_func(request: pytest.FixtureRequest):
        app_path = get_app_path(request, app)
        sa_gen = shiny_app_gen(app_path, timeout_secs=timeout_secs)
>       yield next(sa_gen)
              ^^^^^^^^^^^^

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/pytest/_fixture.py:149: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:341: in shiny_app_gen
    sa = run_shiny_app(
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:269: in run_shiny_app
    sa.wait_until_ready(timeout_secs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <shiny.run._run.ShinyAppProc object at 0x111fe78f0>, timeout_secs = 30

    def wait_until_ready(self, timeout_secs: float) -> None:
        """
        Waits until the shiny app is ready to serve requests.
    
        Parameters
        ----------
        timeout_secs
            The maximum number of seconds to wait for the app to become ready.
    
        Raises
        ------
        ConnectionError
            If there is an error while starting the shiny app.
        TimeoutError
            If the shiny app does not become ready within the specified timeout.
        """
        error_lines: List[str] = []
    
        def stderr_uvicorn(line: str) -> bool:
            error_lines.append(line)
            if "error while attempting to bind on address" in line:
                raise ConnectionError(f"Error while starting shiny app: `{line}`")
            return "Uvicorn running on" in line
    
        if self.stderr.wait_for(stderr_uvicorn, timeout_secs=timeout_secs):
            return
        else:
>           raise TimeoutError(
                "Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:\n"
                + "\n".join(error_lines[-20:])
            )
E           TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/server.py", line 78, in _serve
E           
E               config.load()
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/config.py", line 436, in load
E           
E               self.loaded_app = import_from_string(self.app)
E           
E                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/importer.py", line 19, in import_from_string
E           
E               module = importlib.import_module(module_str)
E           
E                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/importlib/__init__.py", line 90, in import_module
E           
E               return _bootstrap._gcd_import(name[level:], package, level)
E           
E                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
E           
E             File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
E           
E             File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
E           
E             File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
E           
E             File "<frozen importlib._bootstrap_external>", line 999, in exec_module
E           
E             File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
E           
E             File "/Users/alxti/Documents/alex stuff/UBC/block5/532/DSCI-532_2026_10_Salescope/src/app.py", line 17, in <module>
E           
E               from .dflogic import create_summary_table, filter_sales_data
E           
E           ImportError: attempted relative import with no known parent package

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:198: TimeoutError
________________________________________________ ERROR at setup of test_row_dropdown_changes_grouping[chromium] _________________________________________________

request = <SubRequest 'app' for <Function test_initial_kpi_count[chromium]>>

    @pytest.fixture(scope=scope)
    def fixture_func(request: pytest.FixtureRequest):
        app_path = get_app_path(request, app)
        sa_gen = shiny_app_gen(app_path, timeout_secs=timeout_secs)
>       yield next(sa_gen)
              ^^^^^^^^^^^^

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/pytest/_fixture.py:149: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:341: in shiny_app_gen
    sa = run_shiny_app(
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:269: in run_shiny_app
    sa.wait_until_ready(timeout_secs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <shiny.run._run.ShinyAppProc object at 0x111fe78f0>, timeout_secs = 30

    def wait_until_ready(self, timeout_secs: float) -> None:
        """
        Waits until the shiny app is ready to serve requests.
    
        Parameters
        ----------
        timeout_secs
            The maximum number of seconds to wait for the app to become ready.
    
        Raises
        ------
        ConnectionError
            If there is an error while starting the shiny app.
        TimeoutError
            If the shiny app does not become ready within the specified timeout.
        """
        error_lines: List[str] = []
    
        def stderr_uvicorn(line: str) -> bool:
            error_lines.append(line)
            if "error while attempting to bind on address" in line:
                raise ConnectionError(f"Error while starting shiny app: `{line}`")
            return "Uvicorn running on" in line
    
        if self.stderr.wait_for(stderr_uvicorn, timeout_secs=timeout_secs):
            return
        else:
>           raise TimeoutError(
                "Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:\n"
                + "\n".join(error_lines[-20:])
            )
E           TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/server.py", line 78, in _serve
E           
E               config.load()
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/config.py", line 436, in load
E           
E               self.loaded_app = import_from_string(self.app)
E           
E                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/importer.py", line 19, in import_from_string
E           
E               module = importlib.import_module(module_str)
E           
E                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/importlib/__init__.py", line 90, in import_module
E           
E               return _bootstrap._gcd_import(name[level:], package, level)
E           
E                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
E           
E             File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
E           
E             File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
E           
E             File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
E           
E             File "<frozen importlib._bootstrap_external>", line 999, in exec_module
E           
E             File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
E           
E             File "/Users/alxti/Documents/alex stuff/UBC/block5/532/DSCI-532_2026_10_Salescope/src/app.py", line 17, in <module>
E           
E               from .dflogic import create_summary_table, filter_sales_data
E           
E           ImportError: attempted relative import with no known parent package

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:198: TimeoutError
________________________________________________ ERROR at setup of test_reset_button_restores_defaults[chromium] ________________________________________________

request = <SubRequest 'app' for <Function test_initial_kpi_count[chromium]>>

    @pytest.fixture(scope=scope)
    def fixture_func(request: pytest.FixtureRequest):
        app_path = get_app_path(request, app)
        sa_gen = shiny_app_gen(app_path, timeout_secs=timeout_secs)
>       yield next(sa_gen)
              ^^^^^^^^^^^^

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/pytest/_fixture.py:149: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:341: in shiny_app_gen
    sa = run_shiny_app(
/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:269: in run_shiny_app
    sa.wait_until_ready(timeout_secs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <shiny.run._run.ShinyAppProc object at 0x111fe78f0>, timeout_secs = 30

    def wait_until_ready(self, timeout_secs: float) -> None:
        """
        Waits until the shiny app is ready to serve requests.
    
        Parameters
        ----------
        timeout_secs
            The maximum number of seconds to wait for the app to become ready.
    
        Raises
        ------
        ConnectionError
            If there is an error while starting the shiny app.
        TimeoutError
            If the shiny app does not become ready within the specified timeout.
        """
        error_lines: List[str] = []
    
        def stderr_uvicorn(line: str) -> bool:
            error_lines.append(line)
            if "error while attempting to bind on address" in line:
                raise ConnectionError(f"Error while starting shiny app: `{line}`")
            return "Uvicorn running on" in line
    
        if self.stderr.wait_for(stderr_uvicorn, timeout_secs=timeout_secs):
            return
        else:
>           raise TimeoutError(
                "Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:\n"
                + "\n".join(error_lines[-20:])
            )
E           TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/server.py", line 78, in _serve
E           
E               config.load()
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/config.py", line 436, in load
E           
E               self.loaded_app = import_from_string(self.app)
E           
E                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/uvicorn/importer.py", line 19, in import_from_string
E           
E               module = importlib.import_module(module_str)
E           
E                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/importlib/__init__.py", line 90, in import_module
E           
E               return _bootstrap._gcd_import(name[level:], package, level)
E           
E                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           
E             File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
E           
E             File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
E           
E             File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
E           
E             File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
E           
E             File "<frozen importlib._bootstrap_external>", line 999, in exec_module
E           
E             File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
E           
E             File "/Users/alxti/Documents/alex stuff/UBC/block5/532/DSCI-532_2026_10_Salescope/src/app.py", line 17, in <module>
E           
E               from .dflogic import create_summary_table, filter_sales_data
E           
E           ImportError: attempted relative import with no known parent package

/Users/alxti/miniforge3/envs/Salescope/lib/python3.12/site-packages/shiny/run/_run.py:198: TimeoutError
==================================================================== short test summary info ====================================================================
ERROR tests/test_dashboard_playwright.py::test_initial_kpi_count[chromium] - TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
ERROR tests/test_dashboard_playwright.py::test_customer_table_initial_structure[chromium] - TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
ERROR tests/test_dashboard_playwright.py::test_customer_table_initial_cell_values[chromium] - TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
ERROR tests/test_dashboard_playwright.py::test_region_filter_asia_only[chromium] - TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
ERROR tests/test_dashboard_playwright.py::test_purchase_type_filter_two_values[chromium] - TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
ERROR tests/test_dashboard_playwright.py::test_retention_strategy_filter_two_values[chromium] - TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
ERROR tests/test_dashboard_playwright.py::test_row_dropdown_changes_grouping[chromium] - TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
ERROR tests/test_dashboard_playwright.py::test_reset_button_restores_defaults[chromium] - TimeoutError: Shiny app exited without ever becoming ready. Waiting for 'Uvicorn running on' in stderr. Last 20 lines of stderr:
====================================================================== 8 errors in 13.39s =======================================================================
```

--- a/src/app.py
+++ b/src/app.py
@@ -12,6 +12,10 @@ import querychat
 from chatlas import ChatAnthropic, ToolRejectError
 import duckdb
 from pathlib import Path
+import sys
+
+ROOT = Path(__file__).parent
+sys.path.append(str(ROOT))
 
 # src imports
 from dflogic import create_summary_table, filter_sales_data

--- a/src/app.py
+++ b/src/app.py
@@ -14,8 +14,8 @@ import duckdb
 from pathlib import Path
 
 # src imports
-from .dflogic import create_summary_table, filter_sales_data
-from .db import get_base_dataframe, execute_filtered_query
+from dflogic import create_summary_table, filter_sales_data
+from db import get_base_dataframe, execute_filtered_query
 
 
 # used LLM to know how to show actual count/mean inside the box for heatmap

--- a/tests/test_dashboard_playwright.py
+++ b/tests/test_dashboard_playwright.py
@@ -3,7 +3,9 @@
 from shiny.playwright import controller
 from shiny.run import ShinyAppProc
 from shiny.pytest import create_app_fixture
-from playwright.sync_api import Page
+from playwright.sync_api import Page, expect
+
+import re
 
 app = create_app_fixture("../src/app.py")
 
@@ -14,10 +16,13 @@ def test_initial_kpi_count(page: Page, app: ShinyAppProc) -> None:
 
     page.goto(app.url)
     page.wait_for_load_state("networkidle")
+    kpi_locator = page.locator("#kpi_note")
 
-    controller.OutputText(page, "kpi_count").expect_value("7 ⚠️ Low sample", timeout = 10000)
-
-
+    expect(kpi_locator).to_have_text(
+        "⚠️ Low sample size: current filters leave only 7 datapoints.", timeout = 10000
+    )
+    
+    
 def test_customer_table_initial_structure(page: Page, app: ShinyAppProc) -> None:
     
     """Customer Lifetime Value summary table structure, ensuring the KPI table renders correct summary statistics for the selected grouping."""
@@ -27,12 +32,14 @@ def test_customer_table_initial_structure(page: Page, app: ShinyAppProc) -> None
 
     controller.NavPanel(page, id = "advanced_nav", panel_value = "Key Metric Tables").click()
 
+    controller.InputSelect(page, "row_dropdown").set("Region")
+
     customer_df = controller.OutputDataFrame(page, "customer_df")
     customer_df.expect_ncol(6, timeout = 10000)
     customer_df.expect_column_labels(
-        ["Region", "Count", "Mean", "Median", "Maximum", "Total"]
+        ["Group", "Count", "Mean", "Median", "Maximum", "Total"]
     )
-    customer_df.expect_nrow(4)
+    customer_df.expect_nrow(5)
 
 
 def test_customer_table_initial_cell_values(page: Page, app: ShinyAppProc) -> None:
@@ -43,6 +50,8 @@ def test_customer_table_initial_cell_values(page: Page, app: ShinyAppProc) -> No
     page.wait_for_load_state("networkidle")
 
     controller.NavPanel(page, id = "advanced_nav", panel_value = "Key Metric Tables").click()
+
+    controller.InputSelect(page, "row_dropdown").set("Region")
 
     customer_df = controller.OutputDataFrame(page, "customer_df")
     customer_df.expect_cell("Asia", row=0, col=0, timeout = 10000)
@@ -61,12 +70,19 @@ def test_region_filter_asia_only(page: Page, app: ShinyAppProc) -> None:
 
     controller.NavPanel(page, id = "advanced_nav", panel_value = "Key Metric Tables").click()
 
+    controller.InputSelect(page, "row_dropdown").set("Region")
+
     region_checkbox = controller.InputCheckboxGroup(page, "checkbox_group_region")
     region_checkbox.set(["Asia"])
     region_checkbox.expect_selected(["Asia"], timeout = 10000)
 
-    controller.OutputText(page, "kpi_count").expect_value("3 ⚠️ Low sample")
-    controller.OutputDataFrame(page, "customer_df").expect_nrow(1)
+    kpi_locator = page.locator("#kpi_note")
+
+    expect(kpi_locator).to_have_text(
+        "⚠️ Low sample size: current filters leave only 3 datapoints.", timeout = 10000
+    )
+
+    controller.OutputDataFrame(page, "customer_df").expect_nrow(2)
 
 
 def test_purchase_type_filter_two_values(page: Page, app: ShinyAppProc) -> None:
@@ -78,11 +94,17 @@ def test_purchase_type_filter_two_values(page: Page, app: ShinyAppProc) -> None:
 
     controller.NavPanel(page, id = "advanced_nav", panel_value = "Key Metric Tables").click()
 
+    controller.InputSelect(page, "row_dropdown").set("Region")
+
     purchase_checkbox = controller.InputCheckboxGroup(page, "checkbox_group_type")
     purchase_checkbox.set(["Clothing", "Electronics"])
     purchase_checkbox.expect_selected(["Clothing", "Electronics"], timeout = 10000)
 
-    controller.OutputText(page, "kpi_count").expect_value("4 ⚠️ Low sample")
+    kpi_locator = page.locator("#kpi_note")
+
+    expect(kpi_locator).to_have_text(
+        "⚠️ Low sample size: current filters leave only 4 datapoints.", timeout = 10000
+    )
 
 
 def test_retention_strategy_filter_two_values(page: Page, app: ShinyAppProc) -> None:
@@ -92,13 +114,32 @@ def test_retention_strategy_filter_two_values(page: Page, app: ShinyAppProc) -> 
     page.goto(app.url)
     page.wait_for_load_state("networkidle")
 
+
+    # NOTE: These two lines are not actually relevant to this test but removing them causes the following error to show up for some inexplicable reason:
+    """
+     AssertionError: Locator expected to have text '⚠️ Low sample size: current filters leave only 5 datapoints.'
+E       Actual value:  
+E       Call log:
+E         - Expect "to_have_text" with timeout 10000ms
+E         - waiting for locator("#kpi_note")
+E           5 × locator resolved to <div id="kpi_note" aria-live="polite" class="shiny-html-output shiny-bound-output recalculating"></div>
+E             - unexpected value ""
+    """
     controller.NavPanel(page, id = "advanced_nav", panel_value = "Key Metric Tables").click()
+    controller.InputSelect(page, "row_dropdown").set("Region")
 
     strategy_checkbox = controller.InputCheckboxGroup(page, "checkbox_group_strategy")
     strategy_checkbox.set(["Discount", "Email Campaign"])
     strategy_checkbox.expect_selected(["Discount", "Email Campaign"], timeout = 10000)
 
-    controller.OutputText(page, "kpi_count").expect_value("5 ⚠️ Low sample")
+    # there is apparently a race condition here
+    kpi_locator = page.locator("#kpi_note")
+
+
+    expect(kpi_locator).to_have_text(
+        "⚠️ Low sample size: current filters leave only 5 datapoints.", timeout = 10000
+    )
+
 
 
 def test_row_dropdown_changes_grouping(page: Page, app: ShinyAppProc) -> None:
@@ -117,9 +158,9 @@ def test_row_dropdown_changes_grouping(page: Page, app: ShinyAppProc) -> None:
 
     customer_df.expect_ncol(6, timeout = 10000)
     customer_df.expect_column_labels(
-        ["Retention_Strategy", "Count", "Mean", "Median", "Maximum", "Total"]
+        ["Group", "Count", "Mean", "Median", "Maximum", "Total"]
     )
-    customer_df.expect_nrow(3)
+    customer_df.expect_nrow(4)
 
 
 def test_reset_button_restores_defaults(page: Page, app: ShinyAppProc) -> None:
@@ -129,20 +170,42 @@ def test_reset_button_restores_defaults(page: Page, app: ShinyAppProc) -> None:
     page.goto(app.url)
     page.wait_for_load_state("networkidle")
 
+    # NOTE: These two lines are not actually relevant to this test but removing them causes the following error to show up for some inexplicable reason:
+    """
+     AssertionError: Locator expected to have text '⚠️ Low sample size: current filters leave only 5 datapoints.'
+E       Actual value:  
+E       Call log:
+E         - Expect "to_have_text" with timeout 10000ms
+E         - waiting for locator("#kpi_note")
+E           5 × locator resolved to <div id="kpi_note" aria-live="polite" class="shiny-html-output shiny-bound-output recalculating"></div>
+E             - unexpected value ""
+    """
     controller.NavPanel(page, id = "advanced_nav", panel_value = "Key Metric Tables").click()
+    controller.InputSelect(page, "row_dropdown").set("Region")
 
     region_checkbox = controller.InputCheckboxGroup(page, "checkbox_group_region")
     reset_btn = controller.InputActionButton(page, "reset")
-    kpi_count = controller.OutputText(page, "kpi_count")
-
+    
     # changing filter
 
     region_checkbox.set(["Asia"])
-    kpi_count.expect_value("3 ⚠️ Low sample", timeout = 10000)
+    
+    kpi_locator = page.locator("#kpi_note")
+    kpi_locator.wait_for(state="visible")
+
+    expect(kpi_locator).to_have_text(
+        "⚠️ Low sample size: current filters leave only 3 datapoints.", timeout = 10000
+    )
 
     # resetting filter
 
     reset_btn.click()
 
-    kpi_count.expect_value("0 ⚠️ Low sample", timeout = 10000)
+    kpi_locator = page.locator("#kpi_note")
+    expect(page.locator("#kpi_note")).not_to_have_class(re.compile(r"recalculating"), timeout=15000)
+
+    expect(kpi_locator).to_have_text(
+        "⚠️ Low sample size: current filters leave only 7 datapoints.", timeout = 10000
+    )
+
     region_checkbox.expect_selected([])

